File: buzzvil-sdk-ios-v6/buzzvil-sdk-ios-objc/Native/NativeViewController.m
#import <UIKit/UIKit.h>
@import BuzzvilSDK;

@interface NativeViewController : UIViewController

@property (nonatomic, strong) BuzzNative *native;
@property (nonatomic, strong) BuzzNativeAdView *nativeAdView;
@property (nonatomic, strong) BuzzMediaView *mediaView;
@property (nonatomic, strong) UIImageView *iconImageView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *descriptionLabel;
@property (nonatomic, strong) BuzzDefaultCtaView *ctaView;
@property (nonatomic, strong) BuzzNativeViewBinder *viewBinder;
@property (nonatomic, strong) UIActivityIndicatorView *activityIndicatorView;

@end

@implementation NativeViewController

- (instancetype)init {
  self = [super init];
  if (self) {
    _native = [[BuzzNative alloc] initWithUnitId:@"YOUR_NATIVE_UNIT_ID"];
    _nativeAdView = [[BuzzNativeAdView alloc] initWithFrame:CGRectZero];
    _mediaView = [[BuzzMediaView alloc] initWithFrame:CGRectZero];
    _iconImageView = [[UIImageView alloc] initWithFrame:CGRectZero];
    _titleLabel = [[UILabel alloc] initWithFrame:CGRectZero];
    _descriptionLabel = [[UILabel alloc] initWithFrame:CGRectZero];
    _ctaView = [[BuzzDefaultCtaView alloc] initWithFrame:CGRectZero];
    
    _viewBinder = [BuzzNativeViewBinder viewBinderWithBlock:^(BuzzNativeViewBinderBuilder * _Nonnull builder) {
      builder.nativeAdView = _nativeAdView;
      builder.mediaView = _mediaView;
      builder.iconImageView = _iconImageView;
      builder.titleLabel = _titleLabel;
      builder.descriptionLabel = _descriptionLabel;
      builder.ctaView = _ctaView;
    }];
    
    _activityIndicatorView = [[UIActivityIndicatorView alloc] initWithFrame:CGRectZero];
  }
  return self;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self setupView];
  [self setupLayout];
  [self nativeAdLoad];
}

- (void)setupView {
  self.view.backgroundColor = [UIColor systemBackgroundColor];
  [self.view addSubview:self.nativeAdView];
  [self.nativeAdView addSubview:self.mediaView];
  [self.nativeAdView addSubview:self.iconImageView];
  [self.nativeAdView addSubview:self.titleLabel];
  [self.nativeAdView addSubview:self.descriptionLabel];
  [self.nativeAdView addSubview:self.ctaView];
}

- (void)setupLayout {
  self.nativeAdView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.nativeAdView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:16],
    [self.nativeAdView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-16],
    [self.nativeAdView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:8]
  ]];
  
  self.mediaView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.mediaView.topAnchor constraintEqualToAnchor:self.nativeAdView.topAnchor],
    [self.mediaView.leadingAnchor constraintEqualToAnchor:self.nativeAdView.leadingAnchor],
    [self.mediaView.trailingAnchor constraintEqualToAnchor:self.nativeAdView.trailingAnchor],
    [self.mediaView.heightAnchor constraintEqualToAnchor:self.mediaView.widthAnchor multiplier:627.0/1200.0]
  ]];
  
  self.iconImageView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.iconImageView.topAnchor constraintEqualToAnchor:self.mediaView.bottomAnchor constant:8],
    [self.iconImageView.leadingAnchor constraintEqualToAnchor:self.mediaView.leadingAnchor constant:8],
    [self.iconImageView.heightAnchor constraintEqualToConstant:32],
    [self.iconImageView.widthAnchor constraintEqualToConstant:32]
  ]];
  
  self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.titleLabel.centerYAnchor constraintEqualToAnchor:self.iconImageView.centerYAnchor],
    [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.iconImageView.trailingAnchor constant:8],
    [self.titleLabel.trailingAnchor constraintEqualToAnchor:self.nativeAdView.trailingAnchor constant:-8]
  ]];
  
  self.descriptionLabel.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.descriptionLabel.topAnchor constraintEqualToAnchor:self.iconImageView.bottomAnchor constant:8],
    [self.descriptionLabel.leadingAnchor constraintEqualToAnchor:self.iconImageView.leadingAnchor],
    [self.descriptionLabel.trailingAnchor constraintEqualToAnchor:self.titleLabel.trailingAnchor]
  ]];
  
  self.ctaView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.ctaView.topAnchor constraintEqualToAnchor:self.descriptionLabel.bottomAnchor constant:8],
    [self.ctaView.trailingAnchor constraintEqualToAnchor:self.nativeAdView.trailingAnchor constant:-8],
    [self.ctaView.bottomAnchor constraintEqualToAnchor:self.nativeAdView.bottomAnchor constant:-8],
    [self.ctaView.heightAnchor constraintEqualToConstant:32]
  ]];
}

- (void)nativeAdLoad {
  __weak typeof(self) weakSelf = self;
  [self.native subscribeRefreshEventsOnRequest:^{
    // 광고 갱신 할당을 요청한 상태입니다.
  } onSuccess:^(BuzzNativeAd *nativeAd) {
    // 광고 갱신 할당에 성공하면 호출됩니다.
  } onFailure:^(NSError *error) {
    // 더 이상 갱신할 수 있는 광고가 없을 때 호출됩니다.
  }];
  
  [self.native loadOnSuccess:^(BuzzNativeAd *nativeAd) {
    // 초기 광고 할당에 성공하면 호출됩니다.
  } onFailure:^(NSError *error) {
    // 초기 광고 할당에 실패하면 호출됩니다.
  }];
  
  [self.native subscribeAdEventsOnImpressed:^(BuzzNativeAd * _Nonnull) {
    // Native 광고가 유저에게 노출되었을 때 호출됩니다.
  } onClicked:^(BuzzNativeAd * _Nonnull) {
    // 유저가 Native 광고를 클릭했을 때 호출됩니다.
  } onRewardRequested:^(BuzzNativeAd * _Nonnull) {
    // 리워드 적립 요청시에 호출됩니다.
  } onRewarded:^(BuzzNativeAd * _Nonnull, enum BuzzRewardResult) {
    // 리워드 적립 결과를 수신했을 때 호출됩니다.
  } onParticipated:^(BuzzNativeAd * _Nonnull) {
    // 광고 참여가 완료되었을 때 호출됩니다.
  }];
  
  [self.viewBinder bind:self.native];
}

@end

File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-swift/Native/NativeViewController.swift
import BuzzvilSDK
import UIKit

class NativeViewController: UIViewController {
  private lazy var native = BuzzNative(unitId: "YOUR_NATIVE_UNIT_ID")

  private let nativeAdView = BuzzNativeAdView(frame: .zero)
  private let mediaView = BuzzMediaView(frame: .zero)
  private let iconImageView = UIImageView(frame: .zero)
  private let titleLabel = UILabel(frame: .zero)
  private let descriptionLabel = UILabel(frame: .zero)
  private let ctaView = BuzzDefaultCtaView(frame: .zero)
  
  private lazy var benefitHubEntryView: UILabel = {
    let label = UILabel(frame: .zero)
    label.text = "포인트 더 받으러 가기 >"
    
    return label
  }()
  
  private lazy var viewBinder = BuzzNativeViewBinder.Builder()
    .nativeAdView(nativeAdView)
    .mediaView(mediaView)
    .iconImageView(iconImageView)
    .titleLabel(titleLabel)
    .descriptionLabel(descriptionLabel)
    .ctaView(ctaView)
    .build()
  
  private let activityIndicatorView = UIActivityIndicatorView(frame: .zero)
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    setupView()
    setupLayout()
    nativeAdLoad()
  }
  
  private func setupView() {
    // 광고 레이아웃 구성하기
    view.backgroundColor = .systemBackground
    self.view.addSubview(benefitHubEntryView)
    benefitHubEntryView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(showBenefitHub)))
    self.view.addSubview(nativeAdView)
    nativeAdView.addSubview(mediaView)
    nativeAdView.addSubview(iconImageView)
    nativeAdView.addSubview(titleLabel)
    nativeAdView.addSubview(descriptionLabel)
    nativeAdView.addSubview(ctaView)
  }
  
  private func setupLayout() {
    // AutoLayout Constraints 설정
    benefitHubEntryView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      benefitHubEntryView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16),
      benefitHubEntryView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
    ])
    
    nativeAdView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      nativeAdView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
      nativeAdView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16),
      nativeAdView.topAnchor.constraint(equalTo: benefitHubEntryView.bottomAnchor, constant: 8)
    ])
    
    mediaView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      mediaView.topAnchor.constraint(equalTo: nativeAdView.topAnchor),
      mediaView.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor),
      mediaView.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor),
      mediaView.heightAnchor.constraint(equalTo: mediaView.widthAnchor, multiplier: 627/1200),
    ])
    
    iconImageView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      iconImageView.topAnchor.constraint(equalTo: mediaView.bottomAnchor, constant: 8),
      iconImageView.leadingAnchor.constraint(equalTo: mediaView.leadingAnchor, constant: 8),
      iconImageView.heightAnchor.constraint(equalToConstant: 32),
      iconImageView.widthAnchor.constraint(equalToConstant: 32)
    ])
    
    titleLabel.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      titleLabel.centerYAnchor.constraint(equalTo: iconImageView.centerYAnchor),
      titleLabel.leadingAnchor.constraint(equalTo: iconImageView.trailingAnchor, constant: 8),
      titleLabel.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -8)
    ])
    
    descriptionLabel.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      descriptionLabel.topAnchor.constraint(equalTo: iconImageView.bottomAnchor, constant: 8),
      descriptionLabel.leadingAnchor.constraint(equalTo: iconImageView.leadingAnchor),
      descriptionLabel.trailingAnchor.constraint(equalTo: titleLabel.trailingAnchor)
    ])
    
    ctaView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      ctaView.topAnchor.constraint(equalTo: descriptionLabel.bottomAnchor, constant: 8),
      ctaView.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -8),
      ctaView.bottomAnchor.constraint(equalTo: nativeAdView.bottomAnchor, constant: -8),
      ctaView.heightAnchor.constraint(equalToConstant: 32)
    ])
  }
  
  private func nativeAdLoad() {
    // 광고 보여주기 & 광고 이벤트 리스너 등록하기
    // Warning: retain cycle 방지를 위해 closure 내에서 반드시 weak self를 사용해주세요.
    native.subscribeRefreshEvents(
      onRequest: { [weak self] in
        // 광고 갱신 할당을 요청한 상태입니다.
        // 이후에는 onSuccess, onFailure 중 하나가 호출됩니다.
        // 광고 자동 갱신을 시도할 때마다 반복적으로 호출됩니다.
      },
      onSuccess: { [weak self] nativeAd in
        // 광고 갱신 할당에 성공하면 호출됩니다.
      },
      onFailure: { [weak self] error in
        // 더 이상 갱신할 수 있는 광고가 없을 때 호출됩니다.
      }
    )
    
    // 광고 할당을 수행합니다.
    native.load(
      onSuccess: { [weak self] nativeAd in
        // 초기 광고 할당에 성공하면 호출됩니다.
      },
      onFailure: { [weak self] error in
        // 초기 광고 할당에 실패하면 호출됩니다.
      }
    )
    
    // Native 광고 이벤트 처리를 위한 closure를 등록하고 각 이벤트에 따라 필요한 기능을 구현합니다.
    // Warning: retain cycle 방지를 위해 closure 내에서 반드시 weak self를 사용해주세요.
    native.subscribeAdEvents(onImpressed: { [weak self] nativeAd in
      // Native 광고가 유저에게 노출되었을 때 호출됩니다.
    }, onClicked: { [weak self] nativeAd in
      // 유저가 Native 광고를 클릭했을 때 호출됩니다.
    }, onRewardRequested: { [weak self] nativeAd in
      // 리워드 적립 요청시에 호출됩니다.
    }, onRewarded: { [weak self] nativeAd, rewardResult in
      // 리워드 적립 결과를 수신했을 때 호출됩니다.
    }, onParticipated: { [weak self] nativeAd in
      // 광고 참여가 완료되었을 때 호출됩니다.
    })
    
    // 할당된 광고 표시를 자동으로 수행합니다.
    viewBinder.bind(native)
  }
  
  @objc
  private func showBenefitHub() {
    let benefitHub = BuzzBenefitHub()
    
    benefitHub.show(on: self)
  }
}
